{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c99eb63a",
   "metadata": {},
   "source": [
    "# Comparaing simulations against the data from Mitchell\n",
    "Simulated data generated by `simulations.sh parameters.txt`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "013dfa7a-53c6-4347-869c-29f66ee18bc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce19ffa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "from matplotlib.lines import Line2D\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import socket\n",
    "import seaborn as sns\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "\n",
    "from hscpy import mitchell, realisation, parameters\n",
    "from hscpy.figures import PlotOptions, simulations, ToCellFrequency\n",
    "from hscpy.figures import burden as burden_fig\n",
    "from hscpy.figures import sfs as sfs_fig\n",
    "from hscpy import variant\n",
    "\n",
    "from futils import parse_version, snapshot\n",
    "\n",
    "LATEST = True\n",
    "SAVEFIG = True\n",
    "BIGLABELS = False\n",
    "FIGSIZE = [5, 3] if BIGLABELS else [6.4, 4.8]  # default matplotlib\n",
    "EXTENSION = \".svg\"\n",
    "PATH2HSC = Path(\"~\").expanduser() / \"hsc\"\n",
    "PLOT_OPTIONS = PlotOptions(figsize=FIGSIZE, extension=EXTENSION, save=SAVEFIG)\n",
    "NCELLS = 100_000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "985e133e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if socket.gethostname() == \"5X9ZYD3\":\n",
    "    PATH2MITCHELL = Path(\"/mnt/c/Users/terenz01/Documents/SwitchDrive/PhD/hsc\")\n",
    "elif socket.gethostname() == \"LAPTOP-CEKCHJ4C\":\n",
    "    PATH2MITCHELL = Path(\"/mnt/c/Users/fra_t/Documents/PhD/hsc\")\n",
    "else:\n",
    "    PATH2MITCHELL = Path(\"~\").expanduser()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "add2e5a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$PATH2HSC\" --out version\n",
    "$1/target/release/hsc  --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50060586",
   "metadata": {},
   "outputs": [],
   "source": [
    "if LATEST:\n",
    "    VERSION = parse_version(version)\n",
    "else:\n",
    "    VERSION = \"sfs/v3.0.6\"\n",
    "PATH2SAVE = Path(f\"./{VERSION}\")\n",
    "\n",
    "print(\"Running hsc with version:\", VERSION)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "555e7c32",
   "metadata": {},
   "outputs": [],
   "source": [
    "donors = mitchell.donors()\n",
    "donors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e795c3f",
   "metadata": {},
   "source": [
    "## Single-cell mutational burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec108db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# mitchell's donors\n",
    "burden_donors = list()\n",
    "for donor in donors.itertuples():\n",
    "    print(\"loading burden for donor\", donor.name)\n",
    "    burden_donors.append(\n",
    "        mitchell.burden_donor_mitchell(donor.name, donor.age, PATH2MITCHELL, False)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63374b1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DONORS only\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "means, variances = list(), list()\n",
    "for b, c in zip(burden_donors, mcolors.TABLEAU_COLORS.values()):\n",
    "    tot_cells = sum(b[3].values())\n",
    "    assert tot_cells == donors[donors.name == b[0]].cells.iloc[0]\n",
    "    array = snapshot.array_from_hist(b[3])\n",
    "    means.append((b[0], b[1], array.mean()))\n",
    "    variances.append((b[0], b[1], array.var()))\n",
    "\n",
    "    burden_fig.plot_burden(\n",
    "        ax,\n",
    "        b[3],\n",
    "        normalise=True,\n",
    "        options=PLOT_OPTIONS,\n",
    "        ls=\"-\",\n",
    "        marker=\".\",\n",
    "        mew=3,\n",
    "        alpha=0.5,\n",
    "        color=c,\n",
    "        label=f\"{donors.loc[donors.name == b[0], 'age'].iloc[0]} y.o.\",\n",
    "    )\n",
    "ax.legend(fontsize=\"small\", ncols=2)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3d7797f",
   "metadata": {},
   "outputs": [],
   "source": [
    "burden_sims = dict()\n",
    "for donor in donors.itertuples():\n",
    "    burden_sims[donor.name] = realisation.load_all_burden_by_age(\n",
    "        Path(f\"{VERSION}/{donor.cells}cells/burden\")\n",
    "    )[donor.age]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a4605f1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# check that the burden in the sims match the data\n",
    "means_var_s = list()\n",
    "for b in burden_donors:\n",
    "    print(b[0])\n",
    "    fig, ax = plt.subplots(1, 1)\n",
    "    # sims\n",
    "    pooled = snapshot.Uniformise.pooled_distribution(\n",
    "        [bur.burden for bur in burden_sims[b[0]]]\n",
    "    )\n",
    "    m_, v_ = realisation.compute_mean_variance(pooled)\n",
    "    burden_fig.plot_burden(\n",
    "        ax,\n",
    "        pooled,\n",
    "        normalise=False,\n",
    "        color=\"grey\",\n",
    "        marker=\".\",\n",
    "        alpha=0.5,\n",
    "        label=f\"{len(burden_sims[b[0]])} sims\",\n",
    "        options=PLOT_OPTIONS,\n",
    "    )\n",
    "    means_var_s.append((b[0], b[1], m_, v_))\n",
    "    # data\n",
    "    muts, counts = list(b[3].keys()), list(b[3].values())\n",
    "    tot_cells = sum(counts)\n",
    "    assert tot_cells == donors[donors.name == b[0]].cells.iloc[0]\n",
    "    burden_fig.plot_burden(\n",
    "        ax,\n",
    "        b[3],\n",
    "        normalise=True,\n",
    "        color=\"#d95f0e\",\n",
    "        marker=\".\",\n",
    "        bins=10,\n",
    "        alpha=0.5,\n",
    "        label=f\"{b[1]} y.o.\",\n",
    "        options=PLOT_OPTIONS,\n",
    "    )\n",
    "    ax.legend(fontsize=\"small\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "802ec332",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "means_var_df_s = pd.DataFrame(means_var_s, columns=[\"name\", \"age\", \"mean\", \"variance\"])\n",
    "means_df = pd.DataFrame(means, columns=[\"name\", \"age\", \"mean\"])\n",
    "\n",
    "means_df.plot(\n",
    "    ax=ax, x=\"age\", y=\"mean\", marker=\".\", color=\"#d95f0e\", label=\"Mitchell data\"\n",
    ")\n",
    "means_var_df_s.plot(ax=ax, x=\"age\", y=\"mean\", color=\"grey\", label=\"pooled sims\")\n",
    "ax.set_title(\"Mean\")\n",
    "ax.set_xlabel(\"Time [years]\")\n",
    "ax.set_ylabel(\"Single-cell burden\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a44daec",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "variances_df = pd.DataFrame(variances, columns=[\"name\", \"age\", \"variance\"])\n",
    "\n",
    "variances_df.plot(\n",
    "    ax=ax, x=\"age\", y=\"variance\", marker=\".\", color=\"#d95f0e\", label=\"Mitchell data\"\n",
    ")\n",
    "means_var_df_s.plot(ax=ax, x=\"age\", y=\"variance\", color=\"grey\", label=\"pooled sims\")\n",
    "ax.set_title(\"Variance\")\n",
    "ax.set_xlabel(\"Time [years]\")\n",
    "ax.set_ylabel(\"Single-cell burden\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71af1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "m1, m2 = (\n",
    "    snapshot.array_from_hist(burden_donors[0][3]).mean(),\n",
    "    snapshot.array_from_hist(burden_donors[1][3]).mean(),\n",
    ")\n",
    "print(\n",
    "    f\"The mean single-cell mut burden of the two neoborns computed from the genotype matrix is: {m1:.2f}, {m2:.2f}\",\n",
    ")\n",
    "print(\n",
    "    \"from the sims:\", means_var_df_s.loc[means_var_df_s[\"age\"] == 0, \"mean\"].to_numpy()\n",
    ")\n",
    "m1, m2 = (\n",
    "    snapshot.array_from_hist(burden_donors[0][3]).var(),\n",
    "    snapshot.array_from_hist(burden_donors[1][3]).var(),\n",
    ")\n",
    "print(\n",
    "    f\"The variance single-cell mut burden of the two neoborns computed from the genotype matrix is: {m1:.2f}, {m2:.2f}\",\n",
    ")\n",
    "print(\n",
    "    \"from the sims:\",\n",
    "    means_var_df_s.loc[means_var_df_s[\"age\"] == 0, \"variance\"].to_numpy(),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679f0243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# regress neutral donors\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "# neutral donors have no detected exp clone\n",
    "neutral_donors = {\"CB002\", \"KX001\", \"SX001\"}\n",
    "x, y = (\n",
    "    means_df.loc[means_df.name.isin(neutral_donors), \"age\"],\n",
    "    means_df.loc[means_df.name.isin(neutral_donors), \"mean\"],\n",
    ")\n",
    "A = np.vstack([x, np.ones(len(x))]).T\n",
    "m, c = np.linalg.lstsq(A, y, rcond=None)[0]\n",
    "ax.plot(donors.age, m * donors.age + c, \"black\", linewidth=2, linestyle=\"--\")\n",
    "for donor in donors.itertuples():\n",
    "    d_burden = [d for d in burden_donors if d[0] == donor.name][0]\n",
    "    array = snapshot.array_from_hist(d_burden[3])\n",
    "    ax.plot([d_burden[1]] * array.shape[0], array, ls=\"\", marker=\".\", alpha=0.3)\n",
    "    ax.plot([d_burden[1]], array.mean(), ls=\"\", marker=\"x\", mew=2, color=\"black\")\n",
    "ax.plot()\n",
    "print(m, c)\n",
    "ax.set_ylabel(\"Single-cell burden\")\n",
    "ax.set_xlabel(\"Time [years]\")\n",
    "ax.text(x=1, y=1500, s=f\"m={m:.2f}\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65ef951d",
   "metadata": {},
   "source": [
    "## Expanded clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d6debe",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "counts_sims = dict()\n",
    "\n",
    "for donor in donors[[\"name\", \"age\", \"cells\"]].itertuples():\n",
    "    print(\n",
    "        f\"\\tloading sims variant counts for donor {donor.name} with {donor.cells} cells\"\n",
    "    )\n",
    "    counts_sims.update(\n",
    "        variant.load_all_detected_var_counts_by_age(\n",
    "            Path(f\"{VERSION}/{donor.cells}cells/variant_fraction\"), 0.01\n",
    "        )\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2795649",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = variant.variant_counts_detected_df(counts_sims)\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "sns.lineplot(\n",
    "    counts,\n",
    "    x=\"age\",\n",
    "    y=\"variant counts detected\",\n",
    "    errorbar=lambda x: (np.min(x), np.max(x)),\n",
    "    color=\"grey\",\n",
    "    alpha=0.3,\n",
    "    ax=ax,\n",
    "    label=\"min-max\",\n",
    ")\n",
    "ax.plot(donors.age, donors.clones, marker=\".\", color=\"#d95f0e\", label=\"Mitchell\")\n",
    "ax.legend()\n",
    "plt.show()\n",
    "print(counts[[\"variant counts detected\", \"age\"]].groupby(\"age\").describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cd6dbe9",
   "metadata": {},
   "source": [
    "## SFS \n",
    "Combine different data for this plot:\n",
    "1. 1/f^2 sampled prediction (computed here in python)\n",
    "2. Mitchell's SFS (loaded and computed here in python)\n",
    "2. 1/f sampled prediction from Nate's (loaded from external file)\n",
    "3. SFS from simulations (need to generate them)\n",
    "\n",
    "### Generate/load/compute the data\n",
    "####  1. 1/f2 predictions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "621eca78",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# compute the correction for the sims' SFS with sampled distributions from\n",
    "# https://www.biorxiv.org/content/10.1101/2022.11.07.515470v2\n",
    "corrected_variants_one_over_1_squared = dict()\n",
    "for donor in donors.itertuples():\n",
    "    print(\n",
    "        f\"apply sampling correction to SFS of donor {donor.name} with age {donor.age} with sample size {donor.cells}\"\n",
    "    )\n",
    "    corrected_variants_one_over_1_squared[donor.name] = realisation.compute_variants(\n",
    "        realisation.Correction.ONE_OVER_F_SQUARED,\n",
    "        pop_size=NCELLS,\n",
    "        sample_size=donor.cells,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be520211",
   "metadata": {},
   "source": [
    "#### 2. Mitchell's SFS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5bc368e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# there are two donors with the same age 0\n",
    "mitchell_sfs = {\n",
    "    donor.name: mitchell.sfs_donor_mitchell(\n",
    "        donor.name, donor.age, PATH2MITCHELL, remove_indels=False\n",
    "    )\n",
    "    for donor in donors.itertuples()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "635772c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert all(\n",
    "    [\n",
    "        m[2] == donors.loc[donors.name == name, \"cells\"].squeeze()\n",
    "        for name, m in mitchell_sfs.items()\n",
    "    ]\n",
    "), \"number of cells loaded for the SFSF do not match those in donors\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fbaa851",
   "metadata": {},
   "source": [
    "#### 3. 1/f sample predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16009c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# theoretical homeostatic neutral SFS data, from Nate's paper in Elife: for each patient (skipping the neonates)\n",
    "# I evolved until their specific age, and then sampled to the same size as in the data\n",
    "mapping = {\n",
    "    age: f\"predictions_1_over_f/homeostasisSFS_pid{i}.csv\"\n",
    "    for i, age in enumerate(donors.age.unique().tolist()[1:], 3)\n",
    "}\n",
    "mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e54d5664",
   "metadata": {},
   "source": [
    "#### 4. SFS from simulations\n",
    "The data have been generated with the cmd `sfs.sh parameters.txt`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7a4d9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sfs_sims = dict()\n",
    "for donor in donors.itertuples():\n",
    "    sfs_sims[donor.name] = realisation.load_all_sfs_by_age(\n",
    "        Path(f\"{VERSION}/{donor.cells}cells/sfs\")\n",
    "    )[donor.age]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "279641da",
   "metadata": {},
   "source": [
    "#### Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01c4617b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "selected = [\"CB002\", \"KX002\", \"KX008\"]\n",
    "for name in donors.name.unique().tolist():\n",
    "    age = donors.loc[donors.name == name, \"age\"].squeeze()\n",
    "    fig, ax = plt.subplots(1, 1, layout=\"constrained\", figsize=PLOT_OPTIONS.figsize)\n",
    "    sfs_fig.plot_ax_sfs_predictions_data_sims(\n",
    "        ax,\n",
    "        donor=donors[donors.name == name].squeeze(),\n",
    "        corrected_one_over_1_squared=corrected_variants_one_over_1_squared[name],\n",
    "        sfs_sims_donor=None,\n",
    "        mitchell_sfs=mitchell_sfs[name][3],\n",
    "        one_over_f_csv=mapping.get(age),\n",
    "        idx_sim2plot=None,\n",
    "        plot_options=PLOT_OPTIONS,\n",
    "    )\n",
    "\n",
    "    ax.text(\n",
    "        x=0.65,\n",
    "        y=0.9,\n",
    "        s=f\"donor {age} y.o.\",\n",
    "        transform=ax.transAxes,\n",
    "    )\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "    \"\"\"\n",
    "    ax.legend(\n",
    "        fontsize=\"small\",\n",
    "        loc=\"upper right\",\n",
    "        frameon=False,\n",
    "    )\n",
    "    if name not in selected:\n",
    "        ax.legend(\n",
    "            fontsize=\"small\",\n",
    "            loc=\"upper right\",\n",
    "            frameon=False,\n",
    "        )\n",
    "    else:\n",
    "        handles, labels = ax.get_legend_handles_labels()\n",
    "        ax.legend(\n",
    "            handles=[handles[-1]],\n",
    "            labels=[labels[-1]],\n",
    "            fontsize=\"medium\",\n",
    "            loc=\"upper right\",\n",
    "            frameon=False,\n",
    "            handletextpad=0,\n",
    "        )\n",
    "    \"\"\"\n",
    "    if PLOT_OPTIONS.save:\n",
    "        fig.savefig(f\"./sfs_age{age}_{name}{PLOT_OPTIONS.extension}\")\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ad7c90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(10, 1))\n",
    "legend_elements = [\n",
    "    Line2D([1], [1], color=\"black\", alpha=0.8, lw=4, label=\"growth scaling law\"),\n",
    "    Line2D(\n",
    "        [0],\n",
    "        [0],\n",
    "        color=\"black\",\n",
    "        ls=\"--\",\n",
    "        alpha=0.8,\n",
    "        lw=4,\n",
    "        label=\"homeostatsis scaling law\",\n",
    "    ),\n",
    "    Line2D(\n",
    "        [0],\n",
    "        [0],\n",
    "        marker=\"x\",\n",
    "        ls=\"\",\n",
    "        mew=4,\n",
    "        color=\"#d95f0e\",\n",
    "        label=\"data\",\n",
    "        markerfacecolor=\"g\",\n",
    "        markersize=13,\n",
    "    ),\n",
    "]\n",
    "\n",
    "ax.legend(\n",
    "    handles=legend_elements,\n",
    "    mode=\"extend\",\n",
    "    ncols=5,\n",
    "    handletextpad=0.4,\n",
    ")\n",
    "ax.axis(\"off\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0da09d5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "selected = [\"CB002\", \"KX002\", \"KX008\"]\n",
    "for name in donors.name.unique().tolist():\n",
    "    idx_available = [sfs_.parameters.idx for sfs_ in sfs_sims[name]]\n",
    "    print(f\"there are {len(idx_available)} runs for {name}\")\n",
    "    age = donors.loc[donors.name == name, \"age\"].squeeze()\n",
    "    fig, ax = plt.subplots(1, 1, layout=\"constrained\", figsize=PLOT_OPTIONS.figsize)\n",
    "    sfs_fig.plot_ax_sfs_predictions_data_sims(\n",
    "        ax,\n",
    "        donor=donors[donors.name == name].squeeze(),\n",
    "        corrected_one_over_1_squared=corrected_variants_one_over_1_squared[name],\n",
    "        sfs_sims_donor=sfs_sims[name],\n",
    "        mitchell_sfs=mitchell_sfs[name][3],\n",
    "        one_over_f_csv=mapping.get(age),\n",
    "        idx_sim2plot=idx_available[0],\n",
    "        plot_options=PLOT_OPTIONS,\n",
    "    )\n",
    "\n",
    "    ax.text(\n",
    "        x=0.65,\n",
    "        y=0.9,\n",
    "        s=f\"donor {age} y.o.\",\n",
    "        transform=ax.transAxes,\n",
    "    )\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "    \"\"\"\n",
    "    ax.legend(\n",
    "        fontsize=\"small\",\n",
    "        loc=\"upper right\",\n",
    "        frameon=False,\n",
    "    )\n",
    "    if name not in selected:\n",
    "        ax.legend(\n",
    "            fontsize=\"small\",\n",
    "            loc=\"upper right\",\n",
    "            frameon=False,\n",
    "        )\n",
    "    else:\n",
    "        handles, labels = ax.get_legend_handles_labels()\n",
    "        ax.legend(\n",
    "            handles=[handles[-1]],\n",
    "            labels=[labels[-1]],\n",
    "            fontsize=\"medium\",\n",
    "            loc=\"upper right\",\n",
    "            frameon=False,\n",
    "            handletextpad=0,\n",
    "        )\n",
    "    \"\"\"\n",
    "    if PLOT_OPTIONS.save:\n",
    "        fig.savefig(f\"./sfs_age{age}_{name}{PLOT_OPTIONS.extension}\")\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2207fecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(20, 1))\n",
    "legend_elements = [\n",
    "    Line2D([1], [1], color=\"black\", alpha=0.8, lw=4, label=\"growth scaling law\"),\n",
    "    Line2D(\n",
    "        [0],\n",
    "        [0],\n",
    "        color=\"black\",\n",
    "        ls=\"--\",\n",
    "        alpha=0.8,\n",
    "        lw=4,\n",
    "        label=\"homeostatic scaling law\",\n",
    "    ),\n",
    "    Line2D(\n",
    "        [0],\n",
    "        [0],\n",
    "        marker=\"x\",\n",
    "        ls=\"\",\n",
    "        mew=4,\n",
    "        color=\"#d95f0e\",\n",
    "        label=\"data\",\n",
    "        markerfacecolor=\"g\",\n",
    "        markersize=13,\n",
    "    ),\n",
    "    Line2D(\n",
    "        [0],\n",
    "        [0],\n",
    "        marker=\"o\",\n",
    "        ls=\"\",\n",
    "        mew=1,\n",
    "        color=\"grey\",\n",
    "        label=\"single simulation\",\n",
    "        markersize=12,\n",
    "    ),\n",
    "    Line2D([0], [0], color=\"grey\", alpha=0.6, lw=4, label=\"simulation average\"),\n",
    "]\n",
    "\n",
    "ax.legend(\n",
    "    handles=legend_elements,\n",
    "    mode=\"extend\",\n",
    "    ncols=5,\n",
    "    handletextpad=0.4,\n",
    ")\n",
    "ax.axis(\"off\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "211860b0",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for donor in donors.itertuples():\n",
    "    fig, ax = plt.subplots(1, 1, layout=\"tight\")\n",
    "    print(donor.name)\n",
    "\n",
    "    for sfs_s in sfs_sims[donor.name]:\n",
    "        cdf_x_sim, cdf_y_sim = realisation.cdf_from_dict(sfs_s.sfs)\n",
    "        ax.plot(cdf_x_sim / donor.cells, cdf_y_sim, color=\"#bdbdbd\", alpha=0.1)\n",
    "\n",
    "    cdf_x_target, cdf_y_target = realisation.cdf_from_dict(mitchell_sfs[donor.name][3])\n",
    "    ax.plot(cdf_x_target / donor.cells, cdf_y_target, marker=\"o\", color=\"#d95f0e\")\n",
    "    ax.set_xscale(\"log\")\n",
    "    ax.set_ylabel(\"Cumulative distribution\")\n",
    "    ax.set_xlabel(r\"Variant frequency $f$\")\n",
    "    ax.text(\n",
    "        x=0.65,\n",
    "        y=0.1,\n",
    "        s=f\"donor {donor.age} y.o.\",\n",
    "        transform=ax.transAxes,\n",
    "    )\n",
    "    ax.set_ylim([0.9, 1])\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a34db60",
   "metadata": {},
   "source": [
    "### REDO THE SAME PLOT with sims from ABC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8833fa8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx2load = [93820, 577720, 456630, 531350, 589970, 268970, 69560, 262610]\n",
    "# load the data from abc\n",
    "PATH2SIMS = Path(\"/data/scratch\") / f\"hfx923/hsc-draft/v2.2.14\"\n",
    "\n",
    "\n",
    "for r in donors[[\"name\", \"age\", \"cells\"]].itertuples():\n",
    "    path2sfs_abc = Path(PATH2SIMS / f\"{r.cells}cells/sfs/\")\n",
    "    print(f\"\\tloading sims SFS for donor {r.name} with {r.cells} cells\")\n",
    "    sfs_sims_abc = Path(PATH2SIMS / f\"{r.cells}cells/sfs/\")\n",
    "    sfs_sims[r.name].extend(\n",
    "        realisation.load_all_sfs_by_age(path2sfs_abc, idx2load)[r.age]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cdd579c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "selected = [\"CB002\", \"KX002\", \"KX008\"]\n",
    "for i, (name, idx2plot) in enumerate(\n",
    "    zip(donors.name.unique().tolist(), [idx2load[0]] + idx2load)\n",
    "):  # trick for using twice the donor 0\n",
    "    age = donors.loc[donors.name == name, \"age\"].squeeze()\n",
    "    fig, ax = plt.subplots(1, 1, layout=\"constrained\", figsize=PLOT_OPTIONS.figsize)\n",
    "    sfs_fig.plot_ax_sfs_predictions_data_sims(\n",
    "        ax,\n",
    "        donor=donors[donors.name == name].squeeze(),\n",
    "        corrected_one_over_1_squared=corrected_variants_one_over_1_squared[name],\n",
    "        sfs_sims_donor=sfs_sims[name],\n",
    "        mitchell_sfs=mitchell_sfs[name][3],\n",
    "        one_over_f_csv=mapping.get(age),\n",
    "        idx_sim2plot=idx2plot,\n",
    "        plot_options=PLOT_OPTIONS,\n",
    "    )\n",
    "    if i % 3:\n",
    "        ax.set_ylabel(\"\")\n",
    "\n",
    "    ax.text(\n",
    "        x=0.65,\n",
    "        y=0.9,\n",
    "        s=f\"donor {age} y.o.\",\n",
    "        transform=ax.transAxes,\n",
    "    )\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "    \"\"\"\n",
    "    ax.legend(\n",
    "        fontsize=\"small\",\n",
    "        loc=\"upper right\",\n",
    "        frameon=False,\n",
    "    )\n",
    "    if name not in selected:\n",
    "        ax.legend(\n",
    "            fontsize=\"small\",\n",
    "            loc=\"upper right\",\n",
    "            frameon=False,\n",
    "        )\n",
    "    else:\n",
    "        handles, labels = ax.get_legend_handles_labels()\n",
    "        ax.legend(\n",
    "            handles=[handles[-1]],\n",
    "            labels=[labels[-1]],\n",
    "            fontsize=\"medium\",\n",
    "            loc=\"upper right\",\n",
    "            frameon=False,\n",
    "            handletextpad=0,\n",
    "        )\n",
    "    \"\"\"\n",
    "    if PLOT_OPTIONS.save:\n",
    "        fig.savefig(f\"./sfs_age{age}_{name}{PLOT_OPTIONS.extension}\")\n",
    "    fig.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hsc-draft",
   "language": "python",
   "name": "hsc-draft"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
